firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigAutoFetch.java

```diff
@@ -0,0 +1,271 @@
+    // line as partialConfigUpdateMessage and are accumulated together into
+    // currentConfigUpdateMessage.
+    while ((partialConfigUpdateMessage = reader.readLine()) != null) {
+      // Accumulate all the partial parts of the message until we have a full message.
+      currentConfigUpdateMessage += partialConfigUpdateMessage;
+
+      // Closing bracket indicates a full message has just finished.
+      if (partialConfigUpdateMessage.contains("}")) {
+        // Strip beginning and ending of message. If there is not an open and closing bracket,
+        // parseMessage will return an empty message.
+        currentConfigUpdateMessage =
+            parseAndValidateConfigUpdateMessage(currentConfigUpdateMessage);
+        if (!currentConfigUpdateMessage.isEmpty()) {
+          try {
+            JSONObject jsonObject = new JSONObject(currentConfigUpdateMessage);
+
+            if (jsonObject.has(REALTIME_DISABLED_KEY)
+                && jsonObject.getBoolean(REALTIME_DISABLED_KEY)) {
+              retryCallback.onError(
+                  new FirebaseRemoteConfigServerException(
+                      "The server is temporarily unavailable. Try again in a few minutes.",
+                      FirebaseRemoteConfigException.Code.CONFIG_UPDATE_UNAVAILABLE));
+              break;
+            }
+            if (jsonObject.has(TEMPLATE_VERSION_KEY)) {
+              long oldTemplateVersion = configFetchHandler.getTemplateVersionNumber();
+              long targetTemplateVersion = jsonObject.getLong(TEMPLATE_VERSION_KEY);
+              if (targetTemplateVersion > oldTemplateVersion) {
+                autoFetch(MAXIMUM_FETCH_ATTEMPTS, targetTemplateVersion);
+              }
+            }
+          } catch (JSONException ex) {
+            // Message was mangled up and so it was unable to be parsed. User is notified of this
+            // because it there could be a new configuration that needs to be fetched.
+            propagateErrors(
+                new FirebaseRemoteConfigClientException(
+                    "Unable to parse config update message.",
+                    ex.getCause(),
+                    FirebaseRemoteConfigException.Code.CONFIG_UPDATE_MESSAGE_INVALID));
+            Log.e(TAG, "Unable to parse latest config update message.", ex);
+          }
+
+          currentConfigUpdateMessage = "";
+        }
+      }
+    }
+
+    reader.close();
+    inputStream.close();
+  }
+
+  private void autoFetch(int remainingAttempts, long targetVersion) {
+    if (remainingAttempts == 0) {
+      propagateErrors(
+          new FirebaseRemoteConfigServerException(
+              "Unable to fetch the latest version of the template.",
+              FirebaseRemoteConfigException.Code.CONFIG_UPDATE_NOT_FETCHED));
+      return;
+    }
+
+    // Needs fetch to occur between 0 - 4 seconds. Randomize to not cause ddos alerts in backend
```

Comment:
Would leave off the bit about DDOS prevention since that happens on the backend. Could just say "Add 0-4 sec jitter to the next fetch"

Reply:
Updated comment